File: lambda-ecs-durable-python-sam/src/callback_handler.py
import json
import boto3
import os
import hashlib
from aws_durable_execution_sdk_python import (
    DurableContext,
    durable_execution,
)

ecs_client = boto3.client('ecs')

def start_ecs_task_with_callback(cluster, task_definition, subnet1, subnet2, security_group, 
                                  callback_token, message, processing_time):
    """
    Starts an ECS task and passes the callback token via environment variable.
    The ECS task will call Lambda durable execution callback APIs when complete.
    Uses callback_token as idempotency token to prevent duplicate tasks on retry.
    """
    print(f"[CALLBACK] Starting ECS task with callback token")
    
    # Use callback token hash as clientToken for idempotency (max 64 chars)
    client_token = hashlib.sha256(callback_token.encode()).hexdigest()[:64]
    
    response = ecs_client.run_task(
        cluster=cluster,
        taskDefinition=task_definition,
        launchType='FARGATE',
        clientToken=client_token,
        networkConfiguration={
            'awsvpcConfiguration': {
                'subnets': [subnet1, subnet2],
                'securityGroups': [security_group],
                'assignPublicIp': 'ENABLED'
            }
        },
        overrides={
            'containerOverrides': [
                {
                    'name': 'python-callback-container',
                    'environment': [
                        {'name': 'CALLBACK_TOKEN', 'value': callback_token},
                        {'name': 'MESSAGE', 'value': message},
                        {'name': 'PROCESSING_TIME', 'value': str(processing_time)}
                    ]
                }
            ]
        }
    )
    
    if not response['tasks']:
        raise Exception("Failed to start ECS task")
    
    task_arn = response['tasks'][0]['taskArn']
    print(f"[CALLBACK] Task started: {task_arn}")
    
    return task_arn

@durable_execution
def lambda_handler(event, context: DurableContext):
    """
    Lambda durable function that invokes an ECS task and waits for callback.
    
    The ECS task receives a callback token and calls Lambda durable execution
    callback APIs (SendDurableExecutionCallbackSuccess/Failure) when complete.
    
    This function pauses execution while waiting for the callback, with no
    compute charges during the wait period.
    """
    
    # Get configuration from environment variables
    cluster = os.environ['ECS_CLUSTER']
    task_definition = os.environ['TASK_DEFINITION']
    subnet1 = os.environ['SUBNET_1']
    subnet2 = os.environ['SUBNET_2']
    security_group = os.environ['SECURITY_GROUP']
    
    # Get input parameters
    message = event.get('message', 'No message provided')
    processing_time = event.get('processingTime', 5)
    
    try:
        # Create callback to get callback token
        callback = context.create_callback()
        
        print(f"[CALLBACK] Created callback with token: {callback.callback_id[:20]}...")
        
        # Start ECS task with callback token (call directly, no context.step!)
        task_arn = start_ecs_task_with_callback(
            cluster, task_definition, subnet1, subnet2, security_group,
            callback.callback_id, message, processing_time
        )
        
        print(f"[CALLBACK] Waiting for callback from ECS task...")
        
        # Wait for callback (pauses execution here, no compute charges)
        result = callback.result()
        
        print(f"[CALLBACK] Received callback with result")
        
        # Return the result from the callback
        return {
            'statusCode': 200,
            'body': json.dumps({
                'status': 'success',
                'message': 'ECS task completed and sent callback',
                'taskArn': task_arn,
                'result': result
            })
        }
        
    except Exception as e:
        context.logger.error(f"[CALLBACK] Error: {str(e)}")
        
        return {
            'statusCode': 500,
            'body': json.dumps({
                'status': 'error',
                'error': str(e)
            })
        }

File: lambda-durable-functions-nodejs-calling-ecs/src/callback-handler.js
const { ECSClient, RunTaskCommand } = require('@aws-sdk/client-ecs');
const { withDurableExecution } = require('@aws/durable-execution-sdk-js');

const ecs = new ECSClient({ region: process.env.AWS_REGION });

/**
 * Lambda durable function - Callback Pattern
 * 
 * This function demonstrates the callback pattern where:
 * 1. Lambda creates a callback and gets a callback ID
 * 2. Lambda starts an ECS task and passes the callback ID
 * 3. Lambda waits for the callback (suspends without charges)
 * 4. ECS task calls Lambda APIs directly to send success/failure
 * 5. Lambda resumes and returns the result
 */
exports.handler = withDurableExecution(async (event, context) => {
  context.logger.info('Starting Lambda durable function - Callback Pattern');
  context.logger.info('Event:', JSON.stringify(event));

  const message = event.message || 'Hello from Lambda durable function';
  const processingTime = event.processingTime || 10;

  try {
    // Use waitForCallback to create callback and start ECS task
    const result = await context.waitForCallback(
      'ecs-task-callback',
      async (callbackId) => {
        context.logger.info('Callback ID created:', callbackId);
        context.logger.info('Starting ECS task with callback ID...');
        
        // Start ECS task with callback ID as environment variable
        const runTaskParams = {
          cluster: process.env.ECS_CLUSTER,
          taskDefinition: process.env.ECS_TASK_DEFINITION,
          launchType: 'FARGATE',
          networkConfiguration: {
            awsvpcConfiguration: {
              subnets: process.env.ECS_SUBNETS.split(','),
              securityGroups: [process.env.ECS_SECURITY_GROUP],
              assignPublicIp: 'ENABLED'
            }
          },
          overrides: {
            containerOverrides: [
              {
                name: 'worker',
                environment: [
                  { name: 'CALLBACK_ID', value: callbackId },
                  { name: 'MESSAGE', value: message },
                  { name: 'PROCESSING_TIME', value: processingTime.toString() }
                ]
              }
            ]
          }
        };

        const response = await ecs.send(new RunTaskCommand(runTaskParams));
        
        if (!response.tasks || response.tasks.length === 0) {
          throw new Error('Failed to start ECS task');
        }

        const taskArn = response.tasks[0].taskArn;
        context.logger.info('ECS task started:', taskArn);
      },
      {
        timeout: { hours: 1 },  // Wait up to 1 hour for callback
        heartbeatTimeout: { minutes: 5 }  // Expect heartbeat every 5 minutes
      }
    );

    context.logger.info('Callback received with result:', result);

    return {
      statusCode: 200,
      body: JSON.stringify({
        message: 'ECS task completed successfully',
        result: result,
        pattern: 'callback'
      })
    };

  } catch (error) {
    context.logger.error('Error in durable function:', error);
    
    return {
      statusCode: 500,
      body: JSON.stringify({
        message: 'ECS task failed',
        error: error.message,
        pattern: 'callback'
      })
    };
  }
});
